doc: update 1.0 doc outline by adding models and renaming software ar… (#5790)
NVIDIA/TensorRT-LLM: docs/source/1.0/architecture/overview.md

# Architecture Ovewiew

TensorRT-LLM is a toolkit designed to create optimized solutions for Large Language Model (LLM) inference.

> Besides TensorRT, PyTorch can also serve as the backend for TensorRT-LLM. This document provides an overview of the PyTorch Backend architecture.

Reviewer: Still Relevant

## Top Level API

The interface for PyTorch backend is `tensorrt_llm.LLM`.

```python
from tensorrt_llm import LLM
llm = LLM(model=<path_to_llama_from_hf>)
```

The `LLM` also manages the tokenization and detokenization processes of the input.

## PyExecutor

> Similar to the TensorRT backend, which uses [Executor API](../advanced/executor.md), the PyTorch backend employs a `PyExecutor` class.

Reviewer: Since we’re treating the Torch path as the default flow now, I think we should explain everything from scratch here and not assume users know how the TensorRT backend works. We can still mention the TRT equivalent if it helps, but it should be more of a side note rather than the main reference.

```suggestion
The PyExecutor class powers the PyTorch backend and serves as the core execution flow in the Torch backend. It manages request scheduling, model forwarding, and token generation in a modular way. It has a similar interface to the [Executor API](../advanced/executor.md) from the TensorRT backend.
```

Author: Thanks @nvrohanv. I agree with you that we need to make additional modification for adapting torch flow.

However, I think we can address this issue in separate PR since this PR's intention is focus on refine the outline.

What do you think?

Reviewer: I think this one is still relevant since it concerns positioning torch as default.

> This class has a similar interface to Executor, allowing it to be integrated into LLM as an alternative backend.

Reviewer: ```suggestion
```

> Key components of the `PyExecutor` include:

Reviewer: ```suggestion
Key components of `PyExecutor` include:
```

> - Model Engine: Holds the language model and efficiently supports single-step model forward.

Reviewer: Here we say the model engine supports single-step model forward, but later we say the single-step flow involves model forward and running the decoder. I think there is some conflation here between single-step and model forward. Does model forward mean prefill?

> - Decoder: Generates output tokens based on Model Engine outputs. Currently, only greedy search is supported.

Reviewer: ```suggestion
- Decoder: Generates output tokens based on logits output by Model Engine. Currently, only greedy search is supported.
```

> - Scheduler: Decides whether to allocate resources (like KV Cache) for a request and whether to run forward for each request at the current step.

Reviewer: ```suggestion
- Scheduler: Determines which requests should proceed at each step and manages resource allocation (e.g., KV Cache).
```

The single-step flow of PyExecutor involves:

- Fetching new requests from the request queue, if any.
- Scheduling some requests.

> - Running model forward for scheduled requests.

Reviewer: Like I mentioned earlier, saying model forward and then running the decoder seems a bit confusing to me - I believe there will be a model forward for every step (prefill and decode step).

- Running the decoder using the model forward outputs for the scheduled requests.
- Adding output tokens for each request and handling finished requests.

## Model Engine

The core component of `PyExecutor` is the `ModelEngine`, responsible for executing the model's forward pass efficiently on the GPU.
The key method of `ModelEngine` is `forward`, which handles the forward pass computation.
For the PyTorch backend, the derived class is `PyTorchModelEngine`, declared in [pytorch_model_engine.py](../../../tensorrt_llm/_torch/pyexecutor/pytorch_model_engine.py).

## Decoder

> The Decoder generates output tokens based on Model Engine outputs and supports greedy search decoding.

Reviewer: ```suggestion
The Decoder generates output tokens based on Model Engine output logits and supports greedy search decoding.
```

## Scheduler

> The scheduler operates in two steps:

Reviewer: ```suggestion
The scheduler operates in two stages:
```

1. CapacityScheduler: Determines if there are enough resources to accommodate a request.
2. MicroBatchScheduler: Selects some requests for the model to run forward.

Both CapacityScheduler and MicroBatchScheduler currently use C++ bindings.
However, since the interfaces are implemented in Python, customization is possible.
The document [scheduler.md](./scheduler.md) explains how to implement customized scheduling logic.

## ResourceManager

> `ResourceManager` helps allocate and manage these resources that may be needed to run inference for a single request.

Reviewer: ```suggestion
`ResourceManager` helps allocate and manage resources that may be needed to run inference for a single request.
```

It is a container of objects inherited from `BaseResourceManager`, each managing a specific type of resource.
There are three important interfaces for `BaseResourceManager`:

- `prepare_resources`: Called at each step before model forward in PyExecutor for the current batch.
- `update_resources`: Called at each step finish for the current batch.
- `free_resources`: Called at each request finish.

> One crucial resource is the KV Cache for transformer models. The `BaseResourceManager` for KV Cache is `KVCacheManager`.

Reviewer: Can you also provide some other examples of resources "Other resources include ..."
